File: solutions_2453486_0/java/notmichael/tttt.java
import java.util.*;

public class tttt {

	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int cases = Integer.parseInt(in.nextLine());
		for (int i=0; i<cases; i++) {
			char[][] board = new char[4][4];
			for (int j=0; j<4; j++) {
				board[j] = in.nextLine().toCharArray();
			}
			char won = '.';
			boolean fin = false;
			for (int j=0; j<4; j++) {
				int check = 0;
				if (board[j][0]=='T') {
					check = 1;
				}
				if (same(board[j][check],board[j][0]) && same(board[j][check],board[j][1]) && same(board[j][check],board[j][2]) && same(board[j][check],board[j][3])) {
					won = board[j][check];
					fin = true;
					break;
				}
			}
			if (!fin) {
				for (int j=0; j<4; j++) {
					int check = 0;
					if (board[0][j]=='T') {
						check = 1;
					}
					if (same(board[check][j],board[0][j]) && same(board[check][j],board[1][j]) && same(board[check][j],board[2][j]) && same(board[check][j],board[3][j])) {
						won = board[check][j];
						fin = true;
						break;
					}
				}
			}
			if (!fin) {
				for (int j=0; j<2; j++) {
					if (same(board[0][3*j], board[1][1+j]) && same(board[0][3*j], board[2][2-j]) && same(board[0][3*j], board[3][3-3*j]) && same(board[2][2-j], board[3][3-3*j]) && same(board[1][1+j], board[3][3-3*j])) {
						if (board[0][3*j]!='T') {
							won = board[0][3*j];
						}
						else {
							won = board[1][1+j];
						}
						fin = true;
						break;
					}
				}
			}
			if (won == 'X') {
				System.out.println("Case #" + (i+1) + ": X won");
			}
			else if (won == 'O') {
				System.out.println("Case #" + (i+1) + ": O won");
			}
			else {
				boolean done = true;
				for (int j=0; j<4; j++) {
					for (int k=0; k<4; k++) {
						if (board[j][k]=='.') {
							done = false;
						}
					}
				}
				if (done) {
					System.out.println("Case #" + (i+1) + ": Draw");
				}
				else {
					System.out.println("Case #" + (i+1) + ": Game has not completed");
				}
			}
			if(i!=cases-1) {
				in.nextLine();
			}
		}
	}
	
	private static boolean same(char a, char b) {
		if (a==b && a!='.') {
			return true;
		}
		else if (a=='T' && b!='.') {
			return true;
		}
		else if (b=='T' && a!='.') {
			return true;
		}
		else {
			return false;
		}
	}

}

File: solutions_2453486_1/C++/Marshmallow/tttt.cpp

#include <fstream>
#include <iostream>
#include <sstream>
#include "tttt.h"

using namespace std;

void SolveTTTT()
{
	// Set up an input and output file
	ifstream file_in;
	file_in.open("C:\\Users\\John\\Desktop\\CodeJam\\tttt.in");
	ofstream file_out;
	file_out.open("C:\\Users\\John\\Desktop\\CodeJam\\tttt.out");

	// Load the number of test cases
	int test_cases;
	file_in >> test_cases;
	cout << test_cases << " cases" << endl;

	// Each case will have an array of 16 characters
	char game[16];

	// Loop over the cases
	for (int n = 0; n < test_cases; n++)
	{
		// Load the game
		cout << "Board " << (n+1) << ": ";
		for (int i = 0; i < 16; i++)
		{
			file_in >> game[i];
			cout << game[i];
		}
		cout << endl;

		// Now check to see if there's a win, case by case:
		bool finished = true;
		bool Xwon = false;
		bool Owon = false;

		char c;
		for (int i = 0; i < 4; i++)
		{
			// VERTICAL
			c = IsWin(game[i], game[i+4],game[i+8], game[i+12]);
			if (c == '.') finished = false;
			if (c == 'X') Xwon = true;
			if (c == 'O') Owon = true;

			// HORIZONTAL
			c = IsWin(game[4*i + 0], game[4*i + 1],game[4*i + 2],game[4*i + 3]);
			if (c == '.') finished = false;
			if (c == 'X') Xwon = true;
			if (c == 'O') Owon = true;
		}

		// TWO DIAGONALS
		c = IsWin(game[0], game[5], game[10], game[15]);
		if (c == '.') finished = false;
		if (c == 'X') Xwon = true;
		if (c == 'O') Owon = true;

		c = IsWin(game[3], game[6], game[9], game[12]);
		if (c == '.') finished = false;
		if (c == 'X') Xwon = true;
		if (c == 'O') Owon = true;


		file_out << "Case " << (n + 1) << ": ";
		cout << "Case " << (n + 1) << ": ";
		if (Xwon)
		{
			cout << "X won" << endl;
			file_out << "X won" << endl;
			continue;
		}
		if (Owon)
		{
			cout << "O won" << endl;
			file_out << "O won" << endl;
			continue;
		}
		if (finished) { cout << "Draw" << endl; file_out << "Draw" << endl; }
		else { cout << "Game has not completed" << endl; file_out << "Game has not completed" << endl; }
	}

	file_in.close();
	file_out.close();
	return;
}

char IsWin(char a, char b, char c, char d)
{
	bool Xcanwin = true;
	bool Ocanwin = true;

	char group[4] = {a, b, c, d};

	for (int i = 0; i < 4; i++)
	{
		if (group[i] == '.') return '.';
		else if (group[i] == 'O') Xcanwin = false;
		else if (group[i] == 'X') Ocanwin = false;
	}

	if (Xcanwin) return 'X';
	if (Ocanwin) return 'O';

	return 'F';
}

void DealWithChar(char c, bool &finished, bool &Xwon, bool &Owon)
{
	if (c == '.') finished = false;
	if (c == 'X') Xwon = true;
	if (c == 'O') Owon = true;
}




